### PR TITLE
feat: add x/group simulations

### PR DESCRIPTION
**Pull request by aleem1314**
closes: #240

### CODE REVIEW

File: x/group/module/module.go [hunk 1]
```diff
@@ -63,6 +72,29 @@ func (a Module) GetQueryCmd() *cobra.Command {
...
+
+func (a Module) RegisterStoreDecoder(registry sdk.StoreDecoderRegistry) {
+}
+
+func (a Module) WeightedOperations(simState module.SimulationState) []simtypes.WeightedOperation {
+	protoCdc := codec.NewProtoCodec(a.Registry)
+	// TODO: WeightedOperations requires group module query-server
+	return simulation.WeightedOperations(
+		simState.AppParams, simState.Cdc,
+		a.AccountKeeper, a.BankKeeper, protoCdc,
```
Comment (aleem1314) [inline x/group/module/module.go:94]: Is there any way to access group module msgServer here @aaronc?

Comment (aaronc) [reply x/group/module/module.go:94]: We can use the MsgClient with text fixtures. Do you need private state here?

Comment (aleem1314) [reply x/group/module/module.go]: Each simulation operation can depend on the previous one. In the case of the group module, to create a group account 
I don't think . We need a way to pass `group.QueryClient` [here](https://github.com/regen-network/regen-ledger/pull/251/files#diff-e972a3eebdba55ab7bc12235f56652e406f5bc2229c8b4bbe573a2479e9416cdR92).

Comment (aleem1314) [reply x/group/module/module.go]: The whole simulation state is initialized in `app/sim_test.go`. Mostly we are querying the group module state, we need a way to pass `group.QueryClient` from [here](https://github.com/regen-network/regen-ledger/pull/251/files#diff-e972a3eebdba55ab7bc12235f56652e406f5bc2229c8b4bbe573a2479e9416cdR92).
Some context:  Each simulation operation can depend on the previous one. In the case of the group module, to create a group account we need [group-info](https://github.com/regen-network/regen-ledger/pull/251/files#diff-c7b2e4ef3c1b56b3e1be6efbd7d9ed108b8f3457089b7344692be5fe82733608R145) (which is created in the previous operation [SimulateMsgCreateGroup](https://github.com/regen-network/regen-ledger/pull/251/files#diff-c7b2e4ef3c1b56b3e1be6efbd7d9ed108b8f3457089b7344692be5fe82733608R79)).

File: x/group/module/module.go [hunk 2]
```diff
@@ -64,6 +73,27 @@ func (a Module) GetQueryCmd() *cobra.Command {
...
+
+func (a Module) RandomizedParams(r *rand.Rand) []simtypes.ParamChange {
+	return nil
+}
+
+func (a Module) RegisterStoreDecoder(registry sdk.StoreDecoderRegistry) {
+}
+
+func (a Module) WeightedOperations(simState module.SimulationState) []simtypes.WeightedOperation {
+	return simulation.WeightedOperations(
```
Comment (aleem1314) [inline x/group/module/module.go:91]:
```suggestion
        // TODO: simulation.WeightedOperations requires group module's `QueryServer` here.
	return simulation.WeightedOperations(
```

File: app/test_helpers.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+	dbm "github.com/tendermint/tm-db"
+)
+
+// DefaultConsensusParams defines the default Tendermint consensus params used in
+// SimApp testing.
+var DefaultConsensusParams = &abci.ConsensusParams{
```
Comment (blushi) [inline app/test_helpers.go:17]: can't we export this variable directly from the sdk?

File: app/utils.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	simState.ParamChanges = app.SimulationManager().GenerateParamChanges(config.Seed)
+	simState.Contents = app.SimulationManager().GetProposalContents(simState)
+	return app.nm.WeightedOperations(simState)
+}
+
+// CheckExportSimulation exports the app state and simulation parameters to JSON
+// if the export paths are defined.
+func CheckExportSimulation(
```
Comment (blushi) [inline app/utils.go:40]: we could just re-use the `CheckExportSimulation` from the sdk which uses `App` interface, doesn't seem like `RegenApp` is needed here.

File: proto/regen/group/v1alpha1/query.proto [hunk 5]
```diff
@@ -45,6 +48,17 @@ service Query {
...
+message QueryGroupsRequest {
+}
+
+// QueryGroupsResponse is the Query/Groups response type.
+message QueryGroupsResponse {
+
+    // info is the list of GroupInfo for the group.
+    repeated GroupInfo info = 1;
+}
+
```
Comment (blushi) [inline proto/regen/group/v1alpha1/query.proto:61]: we should really add pagination request and response fields.

File: types/module/simulation/simulation.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	"github.com/cosmos/cosmos-sdk/types/simulation"
+	"github.com/regen-network/regen-ledger/types/module/server"
+)
+
+// AppModuleSimulation defines the standard functions that every module should expose
+// for the SDK blockchain simulator
+type AppModuleSimulation interface {
```
Comment (blushi) [inline types/module/simulation/simulation.go:14]: no need to redefine this interface which is already part of the sdk as is.

File: types/module/simulation/simulation.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+	for _, module := range sm.Modules {
+		paramChanges = append(paramChanges, module.RandomizedParams(r)...)
+	}
+
+	return
+}
+
+// WeightedOperations returns all the modules' weighted operations of an application
+func (sm *AppSimulationManager) WeightedOperations(simState module.SimulationState) []simulation.WeightedOperation {
+	// TODO: change it to use New module manager
```
Comment (blushi) [inline types/module/simulation/simulation.go:89]: I guess this can be removed?
```suggestion
```

File: types/module/simulation/simulation.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+// NewAppSimulationManager creates a new SimulationManager object
+//
+// CONTRACT: All the modules provided must be also registered on the module Manager
+func NewAppSimulationManager(serverManager *server.Manager, modules ...module.AppModuleSimulation) *AppSimulationManager {
+	return &AppSimulationManager{
+		Modules:       modules,
+		StoreDecoders: make(sdk.StoreDecoderRegistry),
+		manager:       serverManager,
```
Comment (blushi) [inline types/module/simulation/simulation.go:46]: could we just pass in the result of `serverManager.GetWeightedOperationsHandlers()` instead of the whole `serverManager`?

File: types/module/simulation/simulation.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	return
+}
+
+// WeightedOperations returns all the modules' weighted operations of an application
+func (sm *AppSimulationManager) WeightedOperations(simState module.SimulationState) []simulation.WeightedOperation {
+	// TODO: change it to use New module manager
+	wOps := make([]simulation.WeightedOperation, 0, len(sm.Modules))
+	modules := sm.manager.GetWeightedOperationsHandlers()
```
Comment (blushi) [inline types/module/simulation/simulation.go:91]: In `sm.Modules`, we have the regular modules (from cosmos-sdk) while in `modules`, we'll have the `WeightedOperations` methods from the new regen modules (ecocredit, data, group) so this seems wrong to me.

File: x/group/server/query_server.go [hunk 10]
```diff
@@ -9,6 +9,27 @@ import (
...
+	var infos []*group.GroupInfo
+	var info group.GroupInfo
+	defer it.Close()
+	for {
+		_, err := it.LoadNext(&info)
+		if err != nil {
+			break
+		}
+		infos = append(infos, &info)
+	}
```
Comment (blushi) [inline x/group/server/query_server.go:26]: ditto, use pagination request and `Paginate` function from `orm`.

File: proto/regen/group/v1alpha1/query.proto [hunk 11]
```diff
@@ -45,6 +48,17 @@ service Query {
...
 
+// QueryGroupsRequest is the Query/Groups request type.
+message QueryGroupsRequest {
+}
+
+// QueryGroupsResponse is the Query/Groups response type.
+message QueryGroupsResponse {
+
+    // info is the list of GroupInfo for the group.
+    repeated GroupInfo info = 1;
```
Comment (blushi) [inline proto/regen/group/v1alpha1/query.proto:59]: I'd rather call it this way:
```suggestion
    repeated GroupInfo groups = 1;
```

File: x/group/simulation/operations.go [hunk 12]
```diff
@@ -0,0 +1,748 @@
...
+		_, _, err = app.Deliver(txGen.TxEncoder(), tx)
+		if err != nil {
+			return simtypes.NoOpMsg(group.ModuleName, msg.Type(), "unable to deliver tx"), nil, err
+		}
+
+		return simtypes.NewOperationMsg(&msg, true, ""), nil, err
+	}
+}
+
+// TODO: Add WeightedOperations for MsgVote & MsgExec. Waiting on genesis PR
```
Comment (blushi) [inline x/group/simulation/operations.go:748]: why are you waiting on genesis PR? If it's just about adding `proposal_id` to a `Proposal`, you can go ahead directly here.

**Review by blushi — CHANGES_REQUESTED** [review]
Thanks @aleem1314 for pulling this up!
Generally, I believe we still need to remove some amount of code directly copied from cosmos-sdk.
Also, this doesn't feel right to me to make the bank keeper a dependency of the group server just for simulations, do you see any other possible approach?

**Comment by aleem1314** [comment]
> Also, this doesn't feel right to me to make the bank keeper a dependency of the group server just for simulations, do you see any other possible approach?

I think we can use `bank/QueryClient`, similar to `group/QueryClient`. the issue here is bank module's `QueryClient` doesn't have all the queries, which are required in simulations.

File: x/group/server/query_server.go [hunk 13]
```diff
@@ -9,6 +9,24 @@ import (
 	"github.com/regen-network/regen-ledger/x/group"
 )
 
+func (s serverImpl) Groups(ctx types.Context, request *group.QueryGroupsRequest) (*group.QueryGroupsResponse, error) {
+	it, err := s.groupTable.PrefixScan(ctx, nil, nil)
```
Comment (blushi) [inline x/group/server/query_server.go:13]: You have to use `request.Pagination.Key` in `PrefixScan` `start` parameter. Check out `Paginate` function documentation for more details on how to instantiate iterators to pass into `Paginate`.

File: types/module/simulation/simulation.go [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+	"math/rand"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	"github.com/cosmos/cosmos-sdk/types/simulation"
+)
+
+// AppSimulationManager defines a simulation manager that provides the high level utility
+// for managing and executing simulation functionalities for a group of modules
+type AppSimulationManager struct {
```
Comment (blushi) [inline types/module/simulation/simulation.go:13]: why do we need a different simulation manager than the one already defined in cosmos-sdk?

File: x/group/server/msg_server.go [hunk 15]
```diff
@@ -425,6 +425,7 @@ func (s serverImpl) CreateProposal(ctx types.Context, req *group.MsgCreatePropos
 		return nil, sdkerrors.Wrap(err, "end time conversion")
 	}
 
+	proposalID := s.proposalTable.Sequence().NextVal(ctx)
```
Comment (blushi) [inline x/group/server/msg_server.go:428]: you can just use `s.proposalTable.Sequence().PeekNextVal(ctx)` instead and keep using `s.proposalTable.Create(ctx, m)` which will increment and persist the counter.

File: proto/regen/group/v1alpha1/query.proto [hunk 16]
```diff
@@ -45,6 +48,23 @@ service Query {
   rpc VotesByVoter(QueryVotesByVoterRequest) returns (QueryVotesByVoterResponse);
 }
 
+// QueryGroupsRequest is the Query/Groups request type.
+message QueryGroupsRequest {
+
+  // pagination defines an optional pagination for the request.
+  cosmos.base.query.v1beta1.PageRequest pagination = 2;
```
Comment (blushi) [inline proto/regen/group/v1alpha1/query.proto:55]:
```suggestion
  cosmos.base.query.v1beta1.PageRequest pagination = 1;
```

**Review by blushi — CHANGES_REQUESTED** [review]

File: x/group/simulation/operations.go [hunk 17]
```diff
@@ -0,0 +1,941 @@
...
+
+		_, _, err = app.Deliver(txGen.TxEncoder(), tx)
+		if err != nil {
+			return simtypes.NoOpMsg(group.ModuleName, msg.Type(), "unable to deliver tx"), nil, err
+		}
+		return simtypes.NewOperationMsg(msg, true, ""), nil, err
+	}
+}
+
+func getGroupDetails(sdkCtx sdk.Context, qryClient group.QueryClient, acc simtypes.Account) (groupAdmin string, groupID uint64, op simtypes.OperationMsg, err error) {
```
Comment (blushi) [inline x/group/simulation/operations.go:276]: you could have just used `GroupsByAdmin` query here to query groups by admin account.

**Review by blushi — CHANGES_REQUESTED** [review]

File: types/module/server/manager.go [hunk 18]
```diff
@@ -115,6 +128,22 @@ func (mm *Manager) RegisterModules(modules []module.Module) error {
 	return nil
 }
 
+// WeightedOperations returns all the modules' weighted operations of an application
+func (mm *Manager) WeightedOperations(state sdkmodule.SimulationState, modules []sdkmodule.AppModuleSimulation) []simulation.WeightedOperation {
+	wOps := make([]simulation.WeightedOperation, 0, len(modules))
```
[the remark this answers is not in the document]
Comment (aleem1314) [reply types/module/server/manager.go]: Yeah, I'm getting a lint issue.

[the remark this answers is not in the document]
Comment (blushi) [reply types/module/server/manager.go]: well in this case, let's keep using `make` but with a capacity that makes sense ie `len(modules) + len(mm.weightedOperationsHandlers)` and that will avoid any reallocation.

**Review by blushi — COMMENTED** [review]
@aleem1314 Do we still need `Groups` query?

 Not required. Shall I remove it?

**Comment by blushi** [comment]
> @aleem1314 Do we still need `Groups` query?

If it's not required, let's just remove it.

File: types/module/server/manager.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (blushi) [reply types/module/server/manager.go]: unresolving this as this hasn't been addressed

File: x/group/simulation/genesis.go [hunk 19]
```diff
@@ -0,0 +1,3 @@
+package simulation
+
+// TODO: #270
```
Comment (blushi) [inline x/group/simulation/genesis.go:3]: If genesis PR #270 is merged before this one, then let's do it as part of this PR, else we'll have to create a follow-up issue for adding genesis simulation.

**Review by blushi — COMMENTED** [review]
Looks good overall but I'd like this to be addressed before approving: https://github.com/regen-network/regen-ledger/pull/251#discussion_r593036370

File: types/module/server/manager.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (aleem1314) [reply types/module/server/manager.go]: We cannot use `len(modules) + len(mm.weightedOperationsHandlers)` . Each module returns [`[]simulation.WeightedOperation` ](https://github.com/regen-network/regen-ledger/pull/251/files#diff-a3144c816ac58cb8fbf744d53aa6cc93d3601da586fb23a432db1909b3c4eff8R134-R142).

[the remark this answers is not in the document]
Comment (blushi) [reply types/module/server/manager.go]: oh I see, they might actually be more elements in `wOps` than `len(modules) + len(mm.weightedOperationsHandlers)` but still, pre-allocating would avoid reallocation until we reach `len(modules) + len(mm.weightedOperationsHandlers)` (whereas using `var wOps []simulation.WeightedOperation`, it will realloc from the very first `append`)

File: app/experimental_appconfig.go [hunk 20]
```diff
@@ -79,7 +79,7 @@ func setCustomModules(app *RegenApp, interfaceRegistry types.InterfaceRegistry)
 	newModuleManager := server.NewManager(app.BaseApp, codec.NewProtoCodec(interfaceRegistry))
 
 	// BEGIN HACK: this is a total, ugly hack until x/auth supports ADR 033 or we have a suitable alternative
```
Comment (blushi) [inline app/experimental_appconfig.go:81]: maybe update the comment to explain why we're adding bank keeper too now...

File: x/group/simulation/genesis.go (hunk 19 above)
Comment (blushi) [reply x/group/simulation/genesis.go:3]: now that #270 is merged, could you address this TODO?

File: app/app.go [hunk 21]
```diff
@@ -538,6 +542,11 @@ func (app *RegenApp) SimulationManager() *module.SimulationManager {
 	return app.sm
```
Comment (blushi) [inline app/app.go:542]: do you still need this sim manager `sm *module.SimulationManager`?

**Review by blushi — CHANGES_REQUESTED** [review]

File: app/app.go (hunk 21 above)
Comment (aleem1314) [reply app/app.go:541]: Yes, to initialize SDK modules simulations.

File: app/app.go [hunk 22]
```diff
@@ -538,6 +541,11 @@ func (app *RegenApp) SimulationManager() *module.SimulationManager {
...
 }
 
+// NewManager implements the SimulationApp interface
+func (app *RegenApp) NewManager() *server.Manager {
```
Comment (blushi) [inline app/app.go:545]: why is this needed?

File: app/app.go (hunk 21 above)
Comment (blushi) [reply app/app.go:541]: So we're using the regular sdk `SimulationManager` for all except `WeightedOperation` which needs the new server module manager `smm`, right? Probably better to clarify that in some comments. We will likely need to revisit this as part of #301

**Review by blushi — CHANGES_REQUESTED** [review]

File: app/app.go (hunk 22 above)
Comment (aleem1314) [reply app/app.go]: Removing it. Not using anywhere.

File: x/group/simulation/genesis.go [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+	GroupProposals   = "group-proposals"
+	GroupVote        = "group-vote"
+)
+
+func getGroups(r *rand.Rand, accounts []simtypes.Account) []*group.GroupInfo {
+	groups := make([]*group.GroupInfo, 3)
+	for i := 0; i < 3; i++ {
+		acc, _ := simtypes.RandomAcc(r, accounts)
+		groups[i] = &group.GroupInfo{
+			GroupId:     1,
```
[the remark this answers is not in the document]
Comment (aleem1314) [reply x/group/simulation/genesis.go]: Yeah, make sense.

File: .github/workflows/sims.yml [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+            go.mod
+            go.sum
+      - uses: actions/cache@v2.1.3
+        with:
+          path: ~/go/bin
+          key: ${{ runner.os }}-go-runsim-binary
+        if: env.GIT_DIFF
+      - name: test-test-nondeterminism
+        run: |
+          make sim-regen-nondeterminism
```
Comment (anilCSE) [inline .github/workflows/sims.yml:70]: Should we have two jobs. One for `stable` and one for `experimental` builds?

File: sims.mk [hunk 25]
```diff
@@ -6,40 +6,25 @@
 sim-regen-nondeterminism:
 	@echo "Running nondeterminism test..."
 	@go test -mod=readonly $(APP_DIR) -run TestAppStateDeterminism -Enabled=true \
-		-NumBlocks=100 -BlockSize=200 -Commit=true -Period=0 -v -timeout 24h
+		-NumBlocks=50 -BlockSize=100 -Commit=true -Period=0 -v -timeout 24h -tags=experimental
```
Comment (anilCSE) [inline sims.mk:9]: May be we should have 2 simulations. One for stable as well.

Comment (anilCSE) [reply sims.mk:9]: Also is there any reason for changing the NumBlocks and BlockSize?

File: x/group/simulation/genesis.go [hunk 26]
```diff
@@ -0,0 +1,189 @@
...
+			Metadata:       []byte(simtypes.RandStringOfLength(r, 10)),
+		}
+	}
+	return groupMembers
+}
+
+func getProposals(r *rand.Rand, simState *module.SimulationState) []*group.Proposal {
+	proposals := make([]*group.Proposal, 3)
+	for i := 0; i < 3; i++ {
+
```
Comment (anilCSE) [inline x/group/simulation/genesis.go:79]:
```suggestion

	for i := 0; i < 3; i++ {
```

File: x/group/simulation/genesis.go [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+		from, _ := simtypes.RandomAcc(r, simState.Accounts)
+		to, _ := simtypes.RandomAcc(r, simState.Accounts)
+
+		proposal := &group.Proposal{
+			ProposalId:          uint64(i + 1),
+			Proposers:           []string{simState.Accounts[0].Address.String(), simState.Accounts[1].Address.String()},
```
Comment (anilCSE) [inline x/group/simulation/genesis.go:85]: move this `simState.Accounts[0].Address.String()` and `simState.Accounts[1].Address.String()` out of loop.

File: x/group/simulation/genesis.go [hunk 28]
```diff
@@ -0,0 +1,189 @@
...
+				AbstainCount: "0",
+				VetoCount:    "0",
+			},
+			ExecutorResult: group.ProposalExecutorResultNotRun,
+			Metadata:       []byte(simtypes.RandStringOfLength(r, 50)),
+			SubmittedAt:    gogotypes.Timestamp{Seconds: 1},
+			Timeout:        gogotypes.Timestamp{Seconds: 1000},
+		}
+		err := proposal.SetMsgs([]sdk.Msg{&banktypes.MsgSend{
+			FromAddress: from.Address.String(),
```
Comment (anilCSE) [inline x/group/simulation/genesis.go:103]: `from.Address.String()` is used multiple times. Use a variable.

File: x/group/simulation/genesis.go [hunk 29]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+func getVotes(r *rand.Rand, simState *module.SimulationState) []*group.Vote {
+	votes := make([]*group.Vote, 3)
+
+	for i := 0; i < 3; i++ {
+		votes[i] = &group.Vote{
+			ProposalId:  uint64(i + 1),
+			Voter:       simState.Accounts[i].Address.String(),
+			Choice:      group.Choice_CHOICE_YES,
```
Comment (anilCSE) [inline x/group/simulation/genesis.go:123]: Can we randomize the vote choice or do we need `YES` always?

File: x/group/simulation/genesis.go [hunk 30]
```diff
@@ -0,0 +1,189 @@
...
+	}
+
+	bz, err := simState.Cdc.MarshalJSON(&groupGenesis)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Printf("Selected randomly generated %s parameters:\n%s\n", group.ModuleName, bz)
+	simState.GenState[group.ModuleName] = simState.Cdc.MustMarshalJSON(&groupGenesis)
+
```
Comment (anilCSE) [inline x/group/simulation/genesis.go:188]:
```suggestion
```

File: x/group/simulation/genesis.go [hunk 31]
```diff
@@ -0,0 +1,189 @@
...
+		Proposals:       proposals,
+		Votes:           votes,
+	}
+
+	bz, err := simState.Cdc.MarshalJSON(&groupGenesis)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Printf("Selected randomly generated %s parameters:\n%s\n", group.ModuleName, bz)
```
Comment (anilCSE) [inline x/group/simulation/genesis.go:186]:
```suggestion
	fmt.Printf("selected randomly generated %s parameters:\n%s\n", group.ModuleName, bz)
```

Comment (anilCSE) [reply x/group/simulation/genesis.go:186]: This is genesis state and not just parameters right?
```suggestion
	fmt.Printf("selected randomly generated %s genesis state:\n%s\n", group.ModuleName, bz)
```

File: x/group/simulation/operations.go [hunk 32]
```diff
@@ -0,0 +1,935 @@
...
+	OpMsgCreateProposal                   = "op_weight_msg_create_proposal"
+	OpMsgVote                             = "op_weight_msg_vote"
+	OpMsgExec                             = "ops_weight_msg_exec"
+)
+
+const (
+	WeightCreateGroup    = 150
+	WeightCreateProposal = 100
+	WeightUpdateGroup    = 5
+	WeightGroupMember    = 40
```
Comment (anilCSE) [inline x/group/simulation/operations.go:42]: Any rationale for these weights?

File: x/group/simulation/operations.go [hunk 33]
```diff
@@ -0,0 +1,935 @@
...
+}
+
+// SimulateMsgCreateGroup generates a MsgCreateGroupRequest with random values
+func SimulateMsgCreateGroup(ak exported.AccountKeeper, bk exported.BankKeeper) simtypes.Operation {
+	return func(
+		r *rand.Rand, app *baseapp.BaseApp, ctx sdk.Context, accounts []simtypes.Account, chainID string) (simtypes.OperationMsg, []simtypes.FutureOperation, error) {
+		acc := accounts[0]
+		account := ak.GetAccount(ctx, acc.Address)
+
+		spendableCoins := bk.SpendableCoins(ctx, account.GetAddress())
```
Comment (anilCSE) [inline x/group/simulation/operations.go:174]:
```suggestion
		acc := accounts[0]
		accAddr := acc.Address
		account := ak.GetAccount(ctx, accAddr)

		spendableCoins := bk.SpendableCoins(ctx, accAddr)
```

File: x/group/simulation/operations.go [hunk 34]
```diff
@@ -0,0 +1,935 @@
...
+		fees, err := simtypes.RandomFees(r, ctx, spendableCoins)
+		if err != nil {
+			return simtypes.NoOpMsg(group.ModuleName, group.TypeMsgCreateGroup, "fee error"), nil, err
+		}
+
+		members := []group.Member{
+			{
+				Address:  acc.Address.String(),
```
Comment (anilCSE) [inline x/group/simulation/operations.go:182]:
```suggestion
				Address:  accAddr.String(),
```

or even
```go
accAddrStr := accAddr.String()
```
```suggestion
				Address:  accAddrStr,
```

**Review by anilCSE — COMMENTED** [review]
overall lgtm. A couple of nits and questions

File: sims.mk (hunk 25 above)
Comment (aleem1314) [reply sims.mk]: Updated to original values. I was testing simulations. `-NumBlocks=100` `-BlockSize=200` taking so much time to finish simulation jobs.

File: x/group/simulation/operations.go (hunk 32 above)
Comment (aleem1314) [reply x/group/simulation/operations.go:42]: Yes, If the group or group account modified `MsgVote` & `MsgExec` operations return `noOp`, that's why we have less weight for group & group account update.

Comment (blushi) [reply x/group/simulation/operations.go:42]: could add a comment in the code to explain this then?

**Review by blushi — APPROVED** [review]
LGTM

File: app/experimental_appconfig.go [hunk 35]
```diff
@@ -99,6 +99,7 @@ func setCustomModules(app *RegenApp, interfaceRegistry types.InterfaceRegistry)
 
 	return newModuleManager
 	/* New Module Wiring END */
+	return newModuleManager
```
Comment (atheeshp) [inline app/experimental_appconfig.go:102]:
```
```

Comment (atheeshp) [inline app/experimental_appconfig.go:102]:
```suggestion
```

File: x/group/simulation/operations.go [hunk 36]
```diff
@@ -0,0 +1,938 @@
...
+		if err != nil {
+			return simtypes.NoOpMsg(group.ModuleName, group.TypeMsgVote, "fail to query group info"), nil, err
+		}
+
+		groupAccounts := result.GetGroupAccounts()
+		if len(groupAccounts) == 0 {
+			return simtypes.NoOpMsg(group.ModuleName, group.TypeMsgVote, "no group account found"), nil, nil
+		}
+
+		proposalsResult, err := queryClient.ProposalsByGroupAccount(ctx, &group.QueryProposalsByGroupAccountRequest{Address: groupAccounts[0].Address})
```
Comment (atheeshp) [inline x/group/simulation/operations.go:886]: Can we add helper for `groupAccountAddress`, Same code present in other functions.

**Review by atheeshp — APPROVED** [review]
lgtm overall, nice one @aleem1314, thanks!

**Comment by blushi** [comment]
@anilCSE could you have a final look at this when you have time so we can eventually merge it? Thanks.

File: x/group/module/module.go [hunk 37]
```diff
@@ -71,5 +81,34 @@ func (a Module) RegisterLegacyAminoCodec(cdc *codec.LegacyAmino) {
 }
 
 func (a Module) Route(configurator servermodule.Configurator) sdk.Route {
-	return sdk.NewRoute(group.RouterKey, server.NewHandler(configurator, a.AccountKeeper))
+	return sdk.NewRoute(group.RouterKey, server.NewHandler(configurator, a.AccountKeeper, a.BankKeeper))
+}
+
+// AppModuleSimulation functions
+
+// GenerateGenesisState creates a randomized GenState of the group module.
```
Comment (anilCSE) [inline x/group/module/module.go:89]:
```suggestion
// GenerateGenesisState creates a randomized GenesisState of the group module.
```

File: x/group/simulation/operations.go [hunk 38]
```diff
@@ -0,0 +1,911 @@
...
+	OpMsgExec                             = "ops_weight_msg_exec"
+)
+
+//  If update group or group account txn's executed, `SimulateMsgVote` & `SimulateMsgExec` txn's returns `noOp`.
+//  That's why we have less weight for update group & group-account txn's.
+const (
+	WeightCreateGroup    = 150
+	WeightCreateProposal = 100
+	WeightUpdateGroup    = 5
+	WeightGroupMember    = 40
```
Comment (anilCSE) [inline x/group/simulation/operations.go:44]: We should declare weights for all the sdk messages. The SDK style of doing this is `DefaultWeightMsgMultiSend`, `DefaultWeightMsgDelegate` etc. Also, don't see a reason for keeping `150` for `WeightCreateGroup`. May be it can be just `100`. All sdk weights have either 100 or lesser. May be we should follow the same

**Review by anilCSE — COMMENTED** [review]
implementation lgtm. left a few questions

**Comment by blushi** [comment]
@aleem1314 could you resolve the conflicts so we can merge when you have time?

**Comment by blushi** [comment]
@anilCSE are we good to merge this?

**Comment by aleem1314** [comment]
> @anilCSE are we good to merge this?

Don't merge. One simulation job is failing.

**Comment by blushi** [comment]
> > @anilCSE are we good to merge this?
> 
> Don't merge. One simulation job is failing.

It seems like it's due to group invariants being broken, could you have a look when you have time @aleem1314?

**Comment by aleem1314** [comment]
@anilCSE can you take a final look?

**Review by anilCSE — APPROVED** [review]

**Comment by blushi** [comment]
@aleem1314 are we good to merge this?
